doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+|  Memory Pool with chunk size = 100                     |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
++--------------------------------------------------------+
+```
+
+### Layout With Dynamic Sized Types
+
+With dynamic sized types we move the data part outside of the container and
+add an allocator to the container which allows the container to acquire
+memory for their data part. When we assume that `MyData::image` and
+`MyData::text` require 100 bytes for their data part the memory layout could
+look like this.
+
+```
++--------------------- Memory Manager -------------------+
+|  MemPool with chunk size = 1000                        |
+|  +----------------+----------------+----------------+  |
+|  |  value         |                |                |  |
+|  |  image::mgmt   |                |                |  |
+|  |  text::mgmt    |                |                |  |
+|  +-------------+--+----------------+----------------+  |
+|                |                                       |
+|                +-------------------+-----------+       |
+|  MemPool with chunk size = 100     |           |       |
+|  +---+---+---+---+---+---+---+---+-+-+---+---+-+-+---+ |
+|  |   |   |   |   |   |   |   |   |img|   |   |txt|   | |
+|  |   |   |   |   |   |   |   |   |dat|   |   |dat|   | |
+|  |   |   |   |   |   |   |   |   |   |   |   |   |   | |
+|  +---+---+---+---+---+---+---+---+---+---+---+---+---+ |
++--------------------------------------------------------+
+```
+
+This does not require any adaptions on the subscriber side since the subscriber
+already mapped this memory region which is handled by the memory manager into 
+its local process.
+
+## Design Allocator
+
+The allocator should be designed in a manner so that it can be used inside the
+shared memory within an IPC context. This means the design has to consider that
+vtables, virtual and inheritance are not allowed as well as function pointers
+or the use of `cxx::function` or `cxx::function_ref`.
+
+Therefore this allocator concept is less flexible and more complex as it would
+be when everything would run in the same process since we have to use relative
+pointers and have to allocate everything on the stack.
+
+```
+  +------------------------------+
+  |  AllocatorError [enum class] |
+  |                              |
+  |     # OUT_OF_MEMORY          |
+  |     # MEMORY_SIZE_TOO_LARGE  |
+  +------------------------------+
+
+  +---------------------------------------------------------------+
+  |  Allocator [Concept]                                          |
+  |                                                               |
+  |   - expected<relative_ptr, AllocatorError> allocate(uint64_t) |
```

Comment:
We should not return relaive_ptr or similar. There is no need, we can return raw pointers. Only if we want to store them for IPC use we need to consider fancy pointers. This is not the responsibility of allocate.